docs/proposals/memory-management.md

```diff
@@ -0,0 +1,185 @@
+
+
+#### How is memory managed as of now?
+
+###### Requirements
+
+* Cgroups are used to manage memory on the node.
+* Kernel Out of Memory killer is expected to be enabled. [`/proc/<pid>/oom_score_adj`](https://www.kernel.org/doc/Documentation/filesystems/proc.txt) is used to manage priority for access to memory. The kernel assigns an `oom_score` to individual processes based on the memory usage of the process and a few [other policies](https://lwn.net/Articles/317814/). `oom_score_adj` is taken into account while calculating the oom_score. Higher values for `oom_score_adj` will increase the likelihood for being considered for OOM kills.
+
+###### Node level memory management
+
+Kubelet, the node agent, performs some amount of memory management. Specifically, it can
+
+1. Expose the memory capacity and usage of the node.
+
+2. Apply memory limits to user containers and individual processes on the node.
+
+3. Set up cgroups to track the memory consumption of itself and other system processes like docker daemon, kube-proxy, sshd, etc. The cgroups being setup by kubelet are,
+  a. `/docker-daemon`: The docker daemon is under this cgroup. This cgroup is limited to 70% of the total available memory on the node.
+  b. `/system`: OS processes, excepting init, are under this cgroup.
+  c. `/kubelet`: Kubelet is under this cgroup.
+  d. `/kube-proxy`: The kube-proxy is under this cgroup.
+
+4. Protect system processes from memory pressure on the node:
+   The `oom_score_adj` for kubelet, kube-proxy and other system processes is set to `-999`
+   The `oom_score_adj` for docker daemon is set to `-990`. Docker's memory usage has been unpredicatble in the past and docker daemons restarts are possible under system pressure.
+
+5. Manage Quality of service for user containers: Three classes of containers are currently supported:
+
+   a. `Guaranteed`: Memory is always guaranteed up to the request. These containers will not OOM unless they exceed their request, or system processes consume too much memory, or the kernel overhead is high and the memory in the node is being fully consumed. Note that `request` equals `limit` for this class of containers.
+
+   b. `Burstable`: Memory is guaranteed up until the usage exceeds user specified `request`. These containers do not get killed as soon as they exceed their request. Kernel memory overhead and system processes memory usage will impact memory availability for this class of containers. They are more likely to be killed by the kernel OOM killer once they exceed their request.
```

Comment:
@davidopp - it is set once and never changed from what I could tell right now, but I don't have a running cluster to verify right now.

at start of container:
https://github.com/kubernetes/kubernetes/blob/84f8f01d45f0decc33ad338f3eb8b8a28ce1e3d9/pkg/kubelet/dockertools/manager.go#L1521

Using the following equation:
https://github.com/kubernetes/kubernetes/blob/b661cfd21e9a84560b19ccb68e1b96c87060f1c0/pkg/kubelet/qos/memory_policy.go#L67

I may be misunderstanding the code, but I read it that a containers OOM score adjust value is proportional to the amount of system memory capacity requested.  So if container A requests 10% of machine capacity, it gets a score of 900, but if container B requests 50% of memory capacity, it gets a score of 500.  In a memory pressure situation, it seems to me that even if container B is exceeding its initial request and is using 75% of machine capacity, and container A is actually using less than it requested, that container A is still going to be targeted since its initial request resulted in an inherently higher initial OOM score adjust value.  Right now, we do not appear to distinguish between burstable containers.

this seems like something we should fix unless I am not reading the code correctly.